### ISSUE TITLE
Add support for restricting menu items

### ISSUE THREAD

**Issue by pippinsplugins** [pippinsplugins/Restrict-Content-Pro#202]
This can be done with Widget Login, as shown in [this tutorial](https://pippinsplugins.com/restricting-menu-items-active-members-restrict-content-pro/), but native support would be so much better.

**Comment by chuckreynolds** [comment]
:+1:  but to jump in here. 
I did see the tutorial but I'm not willing to add another plugin into the mix. So native support would be nice. I hard coded `if ( rcp_is_active() ) :` around `wp_nav_menu( $args );` in our theme. That just happened to work for this particular instance. 
Having the ability to just check a box to hide !active on `/wp-admin/nav-menus.php?action=edit&menu=X` would be nice but i'm sure having more customization at that point would be nice for some. Having that field to enter php tags is... well idk but scary to me lol to allow clients access to that post-dev.
So I like having a solution baked in but not sure I'm very excited about that specific option  ;)

**Comment by pippinsplugins** [comment]
@chuckreynolds Why do you refuse to add another plugin? It's a super simple plugin that has as close to zero impact on your site as it can.

I do want to get this in core though :)

**Comment by chuckreynolds** [comment]
@pippinsplugins simple answer. less stuff. I hand these things off and expect people to maintain them. It's something else that'll eventually update.. and/or change something.. may conflict... client(s)/team(s) use it for additional purposes that cause me headaches later.. or causes support questions for me later.. etc. I could go on. Just don't need it unless very necessary. 
I handled it well for the current use. Just thinking of you building it into a public plugin where people have a variety of use cases.

**Comment by pippinsplugins** [comment]
Fair enough. For that particular plugin, however, I'd say you're really safe.

**Comment by pippinsplugins** [comment]
Cool snippet from customer: https://secure.helpscout.net/conversation/133843656/19900/

**Comment by chuckreynolds** [comment]
pretty sure we can't see that @pippinsplugins

**Comment by pippinsplugins** [comment]
Sorry, meant that as an internal note for myself. Here's the code snippet:
```php
if( !is_admin() ) {
	function visibility_check( $items, $menu, $args ) {
		$hidden_items = array();
		foreach( $items as $key => $item ) {
			if( !rcp_user_can_access( get_current_user_id(), $item->object_id ) ) {
				unset( $items[$key] );
				$hidden_items[$item->ID] = '1';
			}
		}

		return $items;
	}
	add_filter( 'wp_get_nav_menu_items', 'visibility_check', 15, 3 );
}
```

**Comment by pippinsplugins** [comment]
We could also extend this to cover complete menus instead of just menu items.

**Comment by pippinsplugins** [comment]
Nice example of a plugin that did this: https://www.cozmoslabs.com/add-ons/navigation-menu-filtering/

**Comment by mindctrl** [comment]
Related: https://core.trac.wordpress.org/ticket/18584

**Comment by mindctrl** [comment]
Gonna work on finishing this up.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by mindctrl** [comment]
I rewrote this yesterday in a way that overrides/extends the `Walker_Nav_Menu_Edit` class. It feels more reliable than doing it with just javascript, and a few other plugins use this method to add custom inputs to the menu edit screen. Until https://core.trac.wordpress.org/ticket/18584 lands in core, I think we should keep this tightened down and as simple as possible, and adopt a more "formal" method when the time comes.

**Comment by pippinsplugins** [comment]
@mindctrl The problem with overriding the walker class is that no other plugin or theme that overwrites the walker can be active at the same time. That's why I went with the JS approach. See `Nav Menu Roles` FAQ for an example list of conflicts: https://wordpress.org/plugins/nav-menu-roles/faq/

**Comment by mindctrl** [comment]
@pippinsplugins I was aware of that and thought it was an acceptable risk. Meaning I figured if someone's using another plugin, or god help a theme, to customize the menu edit screen, they'd need to choose which one was more important to them. On the list of known conflicts there, I don't recall seeing any of those in support.

The code for this to work in javascript felt hacky and inefficient. I wanted it to work and spent quite a bit of time trying to smooth it out before deciding it wasn't an ideal approach. Looping through multiple objects (levels and restrictions) on each menu item to build the interface and match the selections was taking quite some time to run. Another concern was that it wouldn't work at all with any javascript errors happening, which is seemingly pretty frequent.

There isn't an ideal solution until core adds the filter referenced above, so any method we use will be somewhat 'hacky'.

Having said all that, I can backtrack on this method if you disagree with the approach.

**Comment by pippinsplugins** [comment]
Ok let's go ahead with the walker but keeping in mind that we may eventually need to backtrack to the original approach if it raises too many issues in support.

I can see a lot of people already using Nav Menu Roles and then wanting to use our settings on top of it. Also keep in mind that it's not just our settings that might break: other plugins, like Nav Menu Roles, will be broken by RCP.

Two things I want to see on top of this:

1. A help document with an explanation of the conflict and a list of all known conflicts
2. A filter or setting to disable our walker completely

**Comment by mindctrl** [comment]
After talking with Kathy about Nav Menu Roles, we've decided to punt this for now and wait for https://core.trac.wordpress.org/ticket/18584 in WP core. One thing we're missing until that is Customizer support.

If for some reason that ticket doesn't get wrapped up in a reasonable amount of time and people keep asking for this feature, we'll look at tying into Nav Menu Roles to avoid writing a bunch of code that ends up becoming obsolete.

**Comment by nosegraze** [comment]
Just for the record, a customer just submitted a ticket about the "Menu Item Visibility Control" plugin not working and it was because their theme was also overriding the walker class.
